Add thumbdrive & GitHub SCB backup to the "Lightning" section of the guide

VajraOfIndra opened the pull request:
#### What

This PR adds a section in the LND guide to create an automatic thumbdrive-based SCB backup.
As discussed in & Fixes #783 

#### Why

* _Resilience_: Having an updated SCB backup stored outside the SSD is critical to ensure that a user with a bricked node will recover most of his offchain funds (local balances). See discussion in #783. Not having an updated SCB could result in most of the funds locked in the channels for a very long term (or even possibly permanentaly if a peer goes offline permanently).

* _Didactic_: This guide also allow the user to learn how to build and run a shell script from scratch.

#### How

* Added a 'Resilience' section in the LND
* Mentionned an optional remote backup and points to Dropbox bonus guide (the bonus guide will be replaced soon by an alternative using Github, see #783
* Added a small thumbdrive to the list of required equipment in the 'Preparations' page

_Notes/questions:_
* There is probably a lot of room to improve the guide and make it more efficient.
* I'm not the options I chose for the `fstab` file line are the best, could you check and correct?
* I feel the introduction that explains what a SCB is is needed (because of the widespread belief that it aloows the recovery of channels) but I'm sure the wording could be improved a lot

#### Scope

- [x] significant change to core configuration
- [ ] independent bonus guide
- [ ] simple bug fix

Fixes #783 

#### Test & maintenance

_Test:_
* test the guide with a thumbdrive

_Maintenance:_ 
* The link to the Dropbox bonus guide will have to be replaced by a link to the Github bonus guide once this guide is replaced.

#### Animated GIF (optional)

![backup](https://media.giphy.com/media/nK0YXe7YrFNb390uZc/giphy.gif)

Stadicus commented:
I'll review this pr next. It's quite long and I think the LND section would become huge... How about adding a new section to the main guide (after LND, before RTL) called "Channel backup" or similar?

This could be the starting point, but we could also include the GitHub remote backup option later.

Main structure:

* Intro: why are channel backups important?
* TL;DR: options USB and/or Remote
* Guide: USB
* Guide: Remote

WDYT?

Stadicus commented on lnd.md:
```diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  # The script waits for a change in channel.backup. When a change happens (channel opening or closing), a copy of the file is sent to the thumdrive
+  
+  # Location of the channel.backup file used by LND
+  SOURCEFILE=/data/lnd/data/chain/bitcoin/mainnet/channel.backup
```
The main guide stores the file here:

```
/data/lnd-backup/channel.backup
```

VajraOfIndra replied:
I forgot about that.. turns out I comment out the line in my lnd.conf.. I don't remember why though!
As a side note, what is the purpose/benefits of using this dedicated location rather than the default LND one? (as it adds a few steps to the guide)

I update the scb guide to use this dedicated location.

Stadicus replied:
I think I added it as a separate directory so that it's easier to monitor, or even make this directory a GitHub repository. But I'm not sure if we should just leave it in the default location, and copy it with the script to whatever backup location we want to have on top of that...

Stadicus commented on lnd.md:
```diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  SOURCEFILE=/data/lnd/data/chain/bitcoin/mainnet/channel.backup
+  
+  # Location of the backup file in the mounted thumbdrive
+  BACKUPFILE=/thumb/channel.backup
```
Should probably be the following:

```
BACKUPFILE=/mnt/thumbdrive/channel.backup
```

We could also copy the backup file without overwriting the old files (e.g., `channel-20211222-184710.backup` in the following command, or whatever makes sense). 

```
BACKUPFILE="/mnt/thumbdrive/channel-$(date +"%Y%m%d-%H%M%S").backup"
```

VajraOfIndra replied:
Right, that was a leftover from an older version of the guide, fixed!

VajraOfIndra replied:
> We could also copy the backup file without overwriting the old files (e.g., `channel-20211222-184710.backup` in the following command, or whatever makes sense).
> 
> ```
> BACKUPFILE="/mnt/thumbdrive/channel-$(date +"%Y%m%d-%H%M%S").backup"
> ```

Sounds like a great idea, updated

Stadicus commented on lnd.md:
````diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+
+* Paste the following lines. Save and exit.
+
+  ```ini
````
We should add the customary header and the `[Unit]` section to define when the script is started:

```
# RaspiBolt: systemd unit for automatic SCB copy to thumbdrive
# /etc/systemd/system/thumbdrive-scb-backup.service

[Unit]
Description=Thumbdrive SCB Backup daemon
After=lnd.service

```

VajraOfIndra replied:
Fixed!

Stadicus commented on lnd.md:
```diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  RestartSec=1
+  StandardOutput=syslog
+  StandardError=syslog
+  SyslogIdentifier=backup-channels
```
I'd delete the custom "SyslogIdentifier", to keep naming consistent

VajraOfIndra replied:
I deleted it but I don't understand what you mean by 'to keep naming consistent'? do you man we don't do it in the other systemd service units?

Stadicus replied:
I meant that the log entries appear with the name of the systemd unit (e.g., "thumbdrive-scb-backup"). Could be confusing if we introduce a different name for the service.

VajraOfIndra replied:
Ok, thanks a lot for the clarification

Stadicus commented on lnd.md:
```diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  StandardError=syslog
+  SyslogIdentifier=backup-channels
+  User=root
+  Group=root
```
With the changes mentioned above, the service can be executed as user "lnd". No need to configure a specific group.

VajraOfIndra replied:
Ok, changed it to `User=lnd `and remove `Group=root`

Stadicus reviewed (changes requested):
I added some first feedback, no wording or nitpicking just yet :)

TL;DR
* I think we should try to run the service & script with user "lnd" instead of "root" (minimal required permissions)
* This requires the thumbdrive to be mounted as user "lnd"

Also, maybe we could write the backup files in a versioned format, with a filename that includes time & date? This avoids corrupting an existing good backup.

VajraOfIndra commented on lnd.md:
````diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+
+  ```sh
+  $ sudo mkdir /mnt/thumbdrive
+  ```
````
Interesting point! Updated

VajraOfIndra commented on lnd.md:
````diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  $ df -h /dev/sdb
+  > Filesystem      Size  Used Avail Use% Mounted on
+  > /dev/sdb        1.9G  4.0K  1.9G   1% /mnt/thumbdrive
+  ```
````
Ok, fixed

VajraOfIndra commented on lnd.md:
````diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+* Create the mounting directory
+
+  ```sh
+  $ sudo mkdir /mnt/thumbdrive
````
Sure, it's more meaningful, fixed.

VajraOfIndra commented on lnd.md:
````diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+
+  ```sh
+  $ cd /tmp
+  $ nano thumbdrive-scb-backup.sh
````
Good point, it simplifies the section, fixed!

VajraOfIndra commented on lnd.md:
```diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  # Backup function
+  run_backup_on_change () {
+    echo "Copying backup file..."
+    sudo cp $SOURCEFILE $BACKUPFILE
```
Thank you for explaining this, I didn't knwo all these subtleties. I removed the `sudo` from the script

VajraOfIndra replied:
> Maybe it would be cleaner if the script could be run by the lnd user? Then /usr/local/bin is the right location.

So.. we also need to make /mnt/thumbdrive writable by user "lnd", right? what would be the best/most secure way to do this?

Stadicus replied:
> is the gid uid arguments below enough?

Yes, that should do it. A USB thumbdrive formatted with a FAT filesystem does not support permissions on file/directory level, so if we mount the drive with user/group "lnd", that owner automatically owns everything. You can still change things with user "admin" using "sudo", of course.

VajraOfIndra commented on lnd.md:
````diff
@@ -680,6 +680,235 @@ A quick reference with common commands to play around with:
+  ```
+  
+  ```ini
+  UUID=123456 /mnt/thumbdrive vfat auto,noexec,nouser,rw,sync,nosuid,nodev,noatime,nodiratime,nofail 0 0
````
Excellent, updated

VajraOfIndra commented:
> I added some first feedback, no wording or nitpicking just yet :)
> 
> TL;DR
> 
>     * I think we should try to run the service & script with user "lnd" instead of "root" (minimal required permissions)
 
>     * This requires the thumbdrive to be mounted as user "lnd"

> Also, maybe we could write the backup files in a versioned format, with a filename that includes time & date? This avoids corrupting an existing good backup.

Very interesting feedback, thanks! I learnt a lot :) 

I pulled the guide out of LND and placed it between LND and RTL. It feels a bit out of place since the SCB backup is still LND-specific.. but the argument that the LND guide is long enough as it is makes a lot of sense! So I think you're right that it should be its own separate guide (particulalry when the remote GutHub guide will be fleshed out there as well..).

Otherwise, I believe I updated all the suggestions, hoepfully I didn't forget any.. PTAL.

VajraOfIndra commented:
PART 1: Thumbdrive

I reworked the thumbdrive guide a little bit, PTAL

---

PART 2: GitHub

I also created the optional GitHub remote backup guide (replacing the simple pointer to the bonus Dropbox guide that is now deleted in this branch)

However, I encountered some permissions issues I had to work around:
* The `lnd-backup` folder is inside the bitcoin-owned `/data` folder. Therefore cloning the GitHub repo there needs to be done with the bitcoin SSH key..
* But the `lnd-backup` folder itself has to be owned by the "lnd" user for LND to update/write the `channel.backup` file...

To solve the issue, I did the following steps:
* Create SSH keys for the "bitcoin" user to clone the GitHub "lnd-backup" repo
* Change ownership of the "lnd-backup" repo from the "bitcoin" user to the "lnd" user
* Delete the bitcoin SSH deployment key in the Github repo
* Create SSH keys for the "lnd" user and create a deployment key with write access in the GitHub repo

However, it seems convoluted and I'm pretty sure there is a better more straightforward solution to do this... Wdyt?

This led me to wonder about two things:
1. Isn't it a security risk to allow the GitHub repo to have SSH access to the node via the lnd user? Couldn't we create a separate user (e.g. "github") that has only access to a unique folder that only contains a copy of the backup file and nothing else? 
1. Why does /data belongs to the "bitcoin" user as it is also used by other services (Electrs, LND etc)

I let you have a look at this first draft of the remote GitHub backup! I'm sure the script can be improved upon and some steps can be simplify, please amend/slash/correct as much as you want :)

Stadicus reviewed (commented):
First round of review with some questions (e.g., does the separate directory "/data/lnd-backup" make sense?)

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,519 @@
+  ```
+  
+  ```ini
+  UUID=123456 /mnt/thumbdrive-scb vfat auto,noexec,nouser,rw,sync,nosuid,nodev,noatime,nodiratime,nofail,umask=022,gid=1005,uid=1005 0 0
````
Good idea, I changed to XXXX in the relevant places

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+  Restart=always
+  RestartSec=1
+  StandardOutput=syslog
+  StandardError=syslog
```
Ok, good to know. I removed the last 2 lines.

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,519 @@
+  ```
+
+* Start your SSH program (eg. PuTTY) a second time and log in as “admin”. Commands for the second session start with the prompt $2.
+* Simulate a `channel.backup` file change and then exit the session.
````
Great idea. I had no idea what `touch `was doing at the beginning! 
I modified the sentence as such:
`* Simulate a `channel.backup` file change with the `touch` command (that simply updates the timestamp of the file but not its content) and then exit the session.`

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,519 @@
+  > -rwxr-xr-x 1 lnd  lnd     45 Dec 27 17:39 channel-20211227-173403.backup
+  ```
+  
+You're set! Each time you'll open a new channel or close a channel, the backup file in the thumbdrive will be updated.
````
Good point, fixed

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+
+---
+
+## Introduction: why are channel backups important?
```
Fixed

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+
+You can read more about SCBs in [this section of 'Mastering the Lighning Network'](https://github.com/lnbook/lnbook/blob/ec806916edd6f4d1b2f9da2fef08684f80acb671/05_node_operations.asciidoc#node-and-channel-backups){:target="_blank"}.
+
+## TL;DR
```
Agreed, I deleted this heading

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+
+1. Locally, to a USB thumbdrive plugged into the Pi: in case of SSD drive failure only
+1. Remotely, to a GitHub repository: in case of widespread node damage (e.g. flood, fire, etc)
+
```
Well spotted, I missed that one :) fixed!

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+
+## (Optional) Remote backup: GitHub
+
+The thumbdrive-based setup protects the backup from a SSD drive failure. However, it does not protect against a situation where both the SSD drive and USB thumbdrive are destroyed at the same time (*e.g.* fire, food, etc.).  
```
Nope, here food referred to the risk of dropping a Swiss cheese fondue 🧀 on the node and being upset at having lost this delicious meal (and ruined the node as a secondary inconveniance)! ;D

<sub>(fixed!)</sub>

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+
+The thumbdrive-based setup protects the backup from a SSD drive failure. However, it does not protect against a situation where both the SSD drive and USB thumbdrive are destroyed at the same time (*e.g.* fire, food, etc.).  
+
+To protect against this situation, it is necessary to send the backup to a remote location. We will 
```
Deleted "we will"

I added this, please suggest alternative/better wording to justify the choice of GitHub: 

> To protect against this situation, it is necessary to send the backup to a remote location. We will use a private repository on GitHub which is FOSS and free. The SCB file is already encrypted and can therefore be sent to the repository as is and without the need for additional encryption.

Stadicus replied:
Sounds good!

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,519 @@
+
+### Clone the repository to the node
+
+* With the "admin" user, backup  the existing `lnd-backup` folder
```
Agreed, it simplifies the guide quite a bit. I'll be making these changes once I'm back near the node as these are significant changes i need to test live. I'll keep you posted once this is done.

VajraOfIndra replied:
Thinking a bit more about this...there is still the need to create a separate local (on the node) GitHub repository directory... However, if we were to create this directory in the "lnd" user home directory, we could then use only one deployment key and we wouldn't need to first create a pair of keys for the "bitcoin" user! (for now it's needed because /data is owned by "bitcoin")

So, would such a setup be ok?:

- We leave the `lnd.conf` default location for the backup file (i.e. located at `~/.lnd/data/chain/bitcoin/mainnet/channel.backup`)
- The thumbrdive script monitor changes on this file, and copy it to the thumbdrive on changes
- We create a GitHub repository directory at `/home/lnd/lnd-backup` (for this we need to create one pair of keys only, for the user "lnd", no more "bitcoin" user key pair)
- In the GitHub script, we monitor the change of the default LND backup file (i.e. located at `~/.lnd/data/chain/bitcoin/mainnet/channel.backup`). On change, the script copies the file to `/home/lnd/lnd-backup` and then push the changes to the remote repository

Wdyt?

---

On another note... what about simplifying the guide by making only one script for both local (thumdrive) and remote (GitHub) backups? It would remove the optionality of either method.. but it would make the guide significantly smaller in size. Wdyt?

VajraOfIndra commented:
> First round of review with some questions (e.g., does the separate directory "/data/lnd-backup" make sense?)

Thanks! Looking foward to see this section in the guide soon.

For the last comment (the use of the default location), I'll do this when i'll be next to the node as this is a significant changes to a few parts of the guide so I need to test them live. I agree that using the default location is better as it simplifies the guide. I'll keep you posted once this change is completed.

Otherwise, I implemented all the other requested changes, PTAL!

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+
+  ```sh
+  $ sudo mount -a
+  $ df -h thumbdrive-scb
````
Should be `df -h /mnt/thumbdrive-scb`

This is what I get if I run the current command:

```sh
admin@raspibolt3:~ $ df -h thumbdrive-scb
df: thumbdrive-scb: No such file or directory

admin@raspibolt3:~ $ df -h /mnt/thumbdrive-scb
Filesystem      Size  Used Avail Use% Mounted on
/dev/sdb1       7.5G   32K  7.5G   1% /mnt/thumbdrive-scb
```

VajraOfIndra replied:
Well spotted, fixed

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+  > sdb               123456                               vfat     1.9G SCB backup UDisk
+  ```
+
+* Now, get the "lnd" user identifier (UID) and the "lnd" group identifier (GUI) from the `/etc/passwd` database of all user accounts. 
````
"GID" instead of "GUI", sorry, probably my typo

VajraOfIndra replied:
fixed

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+  ```
+
+* Now, get the "lnd" user identifier (UID) and the "lnd" group identifier (GUI) from the `/etc/passwd` database of all user accounts. 
+and copy them into a text editor on your local computer (e.g. here `XXXX` and `XXXX`)
````
`GID` and `UID` are not necessarily the same. Maybe Use `XXXX` and `YYYY`?

VajraOfIndra replied:
I didn't know the numbers could be different. Fixed!

Stadicus reviewed (commented):
Some follow-up nits on PART 1 of the guide :)

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+  ```sh
+  $ cat /home/bitcoin/.ssh/id_rsa.pub
+  > ssh-rsa 1234abcd... bitcoin@raspibolt
+
````
closing backticks are missing

VajraOfIndra replied:
Damn! I regularly forget closing them..! Fixed

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,517 @@
+* Go back to the GitHub repository webpage
+  * Click on "Settings", then "Deploy keys", then "Add deploy keys"
+  * Type a title (e.g., SCB bitcoin user)
+  * In the "Key" box, copy/paste the string generated above starting (e.g. `ssh-rsa 5678efgh... bitcoin@raspibolt`)
```
The "Allow write access" checkbox probably needs to be checked as well:

![image](https://user-images.githubusercontent.com/19550140/151995446-cba34206-38ba-45f0-87c7-913bbf404b91.png)

VajraOfIndra replied:
Done

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+* Clone your newly created empty repository (whem prompted "Are you sure you want to continue connecting", type `yes`)
+
+  ```sh
+  $ git clone git@github.com:VajraOfIndra/lnd-backup.git
````
don't forget to replace "VajraOfIndra" with something like "YourUserName" :)

VajraOfIndra replied:
Well spotted, fixed

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+* Still with user "bitcoin", set up global Git configuration values. The name and email are required but can be dummy values.
+
+  ```sh
+  $ cd /data
````
The `/data` directory is owned by user "bitcoin", which can't clone the repository due to missing SSH keys. Neither can user "admin". And user "lnd" does not have write access here.

I think it's best to clone the repository inside `/data/lnd`: there user "lnd" has write access and it seems fitting.

VajraOfIndra replied:
Done now, no more "bitcoin" user section

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,517 @@
+  > Are you sure you want to continue connecting (yes/no/[fingerprint])? yes
+  > Warning: Permanently added 'github.com,140.82.121.3' (ECDSA) to the list of known hosts.
+  > warning: You appear to have cloned an empty repository.
+  $ cd lnd-backup
```
I'd create a new bullet point here, breaking this code block apart

VajraOfIndra replied:
Done. I also removed most of the ouput to simplify.

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+  > drwxr-xr-x  7 bitcoin bitcoin 4096 Dec 27 18:56 .git
+  ```
+
+* For the backup file to be updated by LND, the folder has to belong to the "lnd" user. With user "admin", change the owner to "lnd"
````
If cloned into `/data/lnd/lnd-remote-backup` (or similar), we can get rid of this whole section

VajraOfIndra replied:
Done

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+  > -rw-r--r--  1 bitcoin bitcoin 11957 Dec 27 19:18 channel.backup
+  ```
+
+#### Allow user "lnd" to access the remote GitHub repo
````
This section can be removed if the repo is originally cloned by user "lnd"

VajraOfIndra replied:
Done

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,517 @@
+
+### Manual backup to remote GitHub directory
+
+#### Move backup file
```
This section needs to be adapted to the new file locations

VajraOfIndra replied:
Done

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+  $ exit
+  ```
+
+### Automatic backup to remote GitHub directory
````
If the original backup monitoring script already includes the conditional code, this section could just be about "enabling" the additional code

VajraOfIndra replied:
Done

Stadicus reviewed (commented):
Went through PART 2 (GitHub backup) of this guide, with the changes we discussed in mind:

* backup is monitored by user "lnd"
* user "lnd" creates SSH keys and clones the repository directly
* the original monitoring script from PART 1 could already include the code for PART 2, but "disabled". Then it just needs to be enabled here.

VajraOfIndra commented:
> Went through PART 2 (GitHub backup) of this guide, with the changes we discussed in mind:
> 
>     * backup is monitored by user "lnd"
> 
>     * user "lnd" creates SSH keys and clones the repository directly
> 
>     * the original monitoring script from PART 1 could already include the code for PART 2, but "disabled". Then it just needs to be enabled here.

I now implemented the new solution we discussed earlier:
* There is now only one script with 2 variables to enable/disable the local/remote backup method
* By default, the script has both enabled (safest option) but users are explained how to disable either one of the 2 methods
* The guide is now much shorter/simpler to use
* The `/data/lnd-backup` directory is now removed from the LND install guide and the backup file locatoon option in the lnd.conf has been removed (the default is now used)
* I added the use of a microSD card as an alternative to a USB stick as it gives users more options and reduce th eneed for them to buy a new piece of hardware.. (but does it have to be mounted as  a USB stick as well? if not, I'll hve to update that section)

I tested the guide on my new node and it works fine. But  will keep an eye until I get a "natural" change of my SCB to see if the backups are made both locally and remotely as intended.

PTAL! Notably:
* is the introduction too long and should we shorten it?
* are the bash script "if" statements ok?

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+  # By default, both methods are used. If you do NOT want to use one of the method, replace "true" by "false" in the two variables below:
+  LOCAL_BACKUP_ENABLED="true"
+  REMOTE_BACKUP_ENABLED="true"
+    
```
Instead of using string variables with values "true" or "false", it would be more correct to use boolean variables. Technically, bash still parses them as strings, but it's more consistent to how other programming languages.

```sh
LOCAL_BACKUP_ENABLED=true
REMOTE_BACKUP_ENABLED=true
```

```sh
if [ "$LOCAL_BACKUP_ENABLED" == true ]; then
  ...
fi
```
https://stackoverflow.com/a/21210966/11424365

Hint: also note that you can combine the `if`/`then` parts on one line.

VajraOfIndra replied:
Great stuff, I amended accordingly.
So `;` not needed if on two lines but needed if `if`/`then `is on one line only.

Stadicus replied:
`;` is basically ending the command, you could have as many commands in one line as you want... they're executed in the given order. In most other cases this is just awful syntax, though.

You can also use `&&`, but this concatenation will abort if a command fails. That's pretty useful, e.g., to print "success" messages, which are automatically skipped if the previous command fails. 

Or the `||` used above to exit the script if `cd` fails.

https://opensource.com/article/18/11/control-operators-bash-shell

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+We recommend to use both methods, but you can choose either one of them, depending on your own requirements and preferences.
+
+---
+
```
I think we should have a note somewhere to check that `lnd.conf` does not contain the `backupfilepath` argument and lnd stores the backup file at the default location.

This is especially important, as existing RaspiBolt v3 setups store the backup file at `/data/lnd-backup`, which now becomes obsolete.

If changes in `lnd.conf` need to be made, lnd then needs to be restarted.

VajraOfIndra replied:
Good point. I added a small "Preparations" section that does that.
Maybe we could delete it down the line once enough time have passed and most people who had set up that option have probably updated thir lnd.conf...

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,390 @@
+* Still with user "lnd", set up global Git configuration values. The name and email are required but can be dummy values.
+
+  ```sh
+  $ cd /data/lnd
````
The `cd` command is not necessary for `git config`. I think it would be better to have it in the next codeblock, just before `git clone`

VajraOfIndra replied:
Ah! I was wondering about this (if I had to be in the same directory or nor.. although I forgot why I though so..).

I really like the suggestion of moving it just before git clone! Fixed (I merged the comment in this now deleted section with the comment of the git clone section that now also contains the two git config commands).

Stadicus replied:
Nice! :+1:

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,390 @@
+  ```sh
+  $ git clone git@github.com:YourUserName/remote-lnd-backup.git
+  > Cloning into 'lnd-backup'...
+  > [...]
````
When cloning the repository, I only get two lines. I think it would be good to include both (it doesn't take more lines) and as one is a "warning", it's good to show that this is okay:

```
Cloning into 'remote-lnd-backup'...
warning: You appear to have cloned an empty repository.
```

VajraOfIndra replied:
I get many more lines... weird. I can't check right now but will get back to you with the output on my node.

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,390 @@
+* Still with user "lnd", enter the Git repository, create a dummy file and push it to your remote GitHub repository 
+
+  ```sh
+  $ cd /data/lnd-backup
````
Path should be updated:

```sh
$ cd cd /data/lnd/remote-lnd-backup
```

VajraOfIndra replied:
well spotted, fixed

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+  $ git add .
+  $ git commit -m "testing"
+  $ git push --set-upstream origin master
+  $ rm test
```
duplicate command (it's deleted in the next code block), should be removed here

RooSoft replied:
yeah I prefer the latter to this one

VajraOfIndra replied:
fixed

Stadicus reviewed (commented):
Testing the whole guide from scratch: it works great! I have some suggestions about structure, the shell script, and some small fixes. I'm not yet including wording suggestions.

![image](https://user-images.githubusercontent.com/19550140/152675624-ee9a847b-b1bb-4869-aab1-c487c9cd180e.png)


Regarding the structure, I'd find it more logical to start with mandatory instructions, e.g., checking that the SCB file is at the expected location, and then installing the script with both backup methods turned off. Then, the user can move on to whatever backup methods they want to enable and drop out of the guide without consequences:

1. Make sure prerequisites hold and install script
2. Optional instructions for USB backup, enabling this part of the script
3. Optional instructions for remote backup, enabling this part of the script.

IMO, the script should have both methods turned off initially, making them "opt-in", and not 'opt-out".

But this structure might just be my personal preference. If others disagree, I'm happy with the guide as is.

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+  * On Windows, follow [this tutorial](https://www.techsolutions.support.com/how-to/how-to-format-a-usb-drive-in-windows-12893){:target="_blank"}
+  
+  * On Mac, follow [this tutorial](https://www.techsolutions.support.com/how-to/how-to-format-a-usb-drive-on-a-mac-12899){:target="_blank"}
+
```
Great suggestion! fixed

VajraOfIndra replied:
(I had done the change but left without saving.. I redid it, it should be visible now)

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+### Formatting
+
+* To ensure that the storage device does not contain malicious code, we will format it on our local computer (select a name easy to recognize, *e.g.*, "SCB backup"). The following guides can be applied to both the thumbdrive and the microSD card.
+
```
Good idea, I amended the sentence, PTAL

VajraOfIndra replied:
(I had done the change but left without saving.. I redid it, it should be visible now)

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,390 @@
+* Create the mounting directory and make it immutable
+
+  ```sh
+  $ sudo mkdir /mnt/storage-device-scb
````
Ok, let's go for `static-channel-backup-external` maybe? I've amended the guide with this new name.

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+  # Local backup function
+  run_local_backup_on_change () {
+    echo "Copying backup file to local storage device..."
+    cp $SOURCEFILE $LOCAL_BACKUP_FILE
```
Nice tip, fixed! (although I don't think I fully grasped the article you just linked to lol)

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,390 @@
+
+  ```sh
+  #!/bin/bash
+  
````
Excellent! I added it with a comment above to explain what it does

Stadicus replied:
Great addition, thanks!

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+  # Remote backup function
+  run_remote_backup_on_change () {
+    echo "Entering Git repository..."
+    cd $GITREPO
```
great, fixed!

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+    echo "Entering Git repository..."
+    cd $GITREPO
+    echo "Making a timestamped copy of channel.backup..."
+    cp $SOURCEFILE $REMOTE_BACKUP_FILE
```
fixed

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,390 @@
+    cp $SOURCEFILE $REMOTE_BACKUP_FILE
+    echo "Committing changes and adding a message"
+    git add .
+    git commit -m "SCB"
```
I tried but couldn't make it work, hence I shamefully opted for no timestamp ;)

Great to have a timestamp there as well, thanks! fixed

Stadicus replied:
Nice. It's a bit redundant, but at least now the git commit history includes this information as well:

![image](https://user-images.githubusercontent.com/19550140/152686393-f4472306-6c79-4ef2-b6a7-17c7eab4b25a.png)

VajraOfIndra replied:
Oh, I didn't know about that command, nice!

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,390 @@
+  ```
+
+* Check the following lines of code and paste them in nano. By default, both local and remote backup methods are enabled in the script. If you do not plan to use one of the method, change the corresponding variable (`LOCAL_BACKUP_ENABLED` or `REMOTE_BACKUP_ENABLED`) to "false". Save and exit.
+
````
* I broke up some of the long lines. Do you also split a command line if greater than 80 characters: e.g. how should be handle: `LOCAL_BACKUP_FILE="/mnt/static-channel-backup-external/channel-$(date +"%Y%m%d-%H%M%S").backup"`
* I removed the last blanks
* Nice tip the shellcheck webite! I tried it and it returned an OK. I'll use it from now on, thanks a lot

Stadicus replied:
> Do you also split a command line

That's possible using `\` at the end of the line, but I usually don't do that

VajraOfIndra replied:
> > Do you also split a command line
> 
> That's possible using `\` at the end of the line, but I usually don't do that

Ok, good to know

RooSoft commented on static-channel-backup.md:
```diff
@@ -0,0 +1,389 @@
+
+You can read more about SCBs in [this section of 'Mastering the Lighning Network'](https://github.com/lnbook/lnbook/blob/ec806916edd6f4d1b2f9da2fef08684f80acb671/05_node_operations.asciidoc#node-and-channel-backups){:target="_blank"}.
+
+The guide will show how to set up an automatic Static Channel Backup:
```
Might be a good idea to mention that those are two mutually exclusive solutions... people can implement either one or both of them. Reading this leaves me the impression that they are both necessary steps.

RooSoft replied:
Also, maybe change numbers with bullet points so it doesn't convey the idea of a hierarchy or something

RooSoft commented on static-channel-backup.md:
````diff
@@ -0,0 +1,389 @@
+* Go back to the SSH session, delete the test file, and exit the "lnd" user
+
+  ```sh
+  $ rm test
````
Might be a good idea to commit that change and push it to github so the repo is in sync with the clone. I'd replace the `rm test` line with all of following:

```bash
git rm test
git add .
git commit -m 'removing test file'
git push
```

VajraOfIndra replied:
Good idea. Amended acordingly.

RooSoft replied:
Unsure what it adds, but `git rm test` would ask git to remove the file itself... still removing the file using bash directly also is fine.

RooSoft commented on static-channel-backup.md:
````diff
@@ -0,0 +1,388 @@
+  > [...]
+  ```
+
+* Check the content of your local storage device. It should now contain a backup file with the date/time corresponding to the test made just above
````
The following won't apply as is when people choose remote only... The following bullet point also won't apply for local only. I'd maybe make a single step for both of them, and sub-steps following an explanation that this is a checkup based on previous choices.

VajraOfIndra replied:
Good point. I simpy added "If you enabled local backup, " and "If you enabled remote backup," at the start of each lines. PTAL, wdyt?

RooSoft replied:
That's way better!

RooSoft reviewed (changes requested):
Great work, some minor things to take in consideration...

VajraOfIndra commented on bonus/lightning/index.md:
```diff
@@ -17,11 +17,6 @@ has_toc: false
 ---
 
-## Resilience
-* **[Automatic Static Channel Backup (SCB) to Dropbox](static-backup-dropbox.md)** - (for RaspiBolt v1 & v2) have an updated SCB copy in case you brick your node and want to recover your funds onchain
```
I deleted this section from the index file.. what do you mean it leads to a 404?

RooSoft replied:
Think I was looking at something outdated, it's ok now!

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,388 @@
+* Still with user "lnd", enter the Git repository, create a dummy file and push it to your remote GitHub repository 
+
+  ```sh
+  $ cd /data/remote-lnd-backup
````
I missed this before: should be

```sh
$ cd /data/lnd/remote-lnd-backup
```

VajraOfIndra commented on preparations.md:
```diff
@@ -36,6 +36,7 @@ You need the following hardware:
 * Raspberry Pi 4, with 4+ GB RAM
 * official Raspberry Pi power adapter: other adapters are a common cause for reliability issues
 * external storage: 1+ TB with USB3, an SSD is recommended
+* external storage: a USB thumbdrive (even the smallest size thumdrive will do)
```
Maybe we could say:

(Optional) external backup device: a small USB thumbdrive or microSD card

wdyt?

RooSoft replied:
At the very least, I'd explain why they're optional and what they might be for.... but I'd even separate them in a optional hardware list just below, so that the mandatory stuff list seems lighter at first sight to newcomers. 

But you know, that's ultra minor, just adding thoughts here.

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,517 @@
+
+---
+
+## Local backup: USB thumbdrive
```
This is an outdated title. The new title is "Local backup preparations" which is device agnostic. The intro to the section reminds the user that it is an optional section.

Yes, the new version of the guide make either option optional, none is preferred or recomended (well, using both is recommended, but not using one rather than the other is what I meant)

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,517 @@
+
+Static Channels Backup is a feature of LND that allows for the onchain recovery of lightning channel balances in the case of a bricked node. Despite its name, it does not allow the recovery of your LN channels but simply increases the chance that you'll recover all (or most) of your offchain (local) balances.  
+
+The SCB contains all the necessary channel information used for the recovery process which is called the Data Loss Protection (DLP). It is a foolproof safe backup mechanism (*i.e.*, there is no risk of penalty transactions being triggered which would result in the entire local). During recovery, the SCB is used by LND to know who were you peers with whom you had channels. LND send all your online ...
```
well spotted, fixed!

VajraOfIndra commented on static-channel-backup.md:
```diff
@@ -0,0 +1,517 @@
+
+This is why it is recommended to set up an automatic SCB update mechanism that:
+
+1. Create a new (or update an existing) SCB file each time you open a channel (and close, although this is less important)
```
Isn't that for the main `channel.db` database though? The `channel.backup` SCB file is only updated when a channel is opened or closed afaik.

See here: https://github.com/lightningnetwork/lnd/releases/tag/v0.6-beta

> We call these “static” backups, as they only need to be obtained once for a given channel and are valid until the channel has been closed

RooSoft replied:
Looks like I've learnt something here... I was just plain wrong on that, remark, sorry for that!

VajraOfIndra commented on static-channel-backup.md:
````diff
@@ -0,0 +1,517 @@
+* Create a new shell script file
+
+  ```sh
+  $ sudo nano /usr/local/bin/thumbdrive-scb-backup.sh
````
Glad to learn this. I removed the .sh endinf. Note that the new script name is also much simpler now: `scb-backup`

VajraOfIndra commented:
> Testing the whole guide from scratch: it works great! I have some suggestions about structure, the shell script, and some small fixes. I'm not yet including wording suggestions.

So I went through all your and @RooSoft comments. I think I didn't miss any.. although hard to keep track of (particulalry with GitHub sometimes collapsing comments without any reason!).

> Regarding the structure, I'd find it more logical to start with mandatory instructions, e.g., checking that the SCB file is at the expected location, and then installing the script with both backup methods turned off. Then, the user can move on to whatever backup methods they want to enable and drop out of the guide without consequences:
> 
>     1. Make sure prerequisites hold and install script
> 
>     2. Optional instructions for USB backup, enabling this part of the script
> 
>     3. Optional instructions for remote backup, enabling this part of the script.
> 
> 
> IMO, the script should have both methods turned off initially, making them "opt-in", and not 'opt-out".
> 
> But this structure might just be my personal preference. If others disagree, I'm happy with the guide as is.

I like this solutiona as well, yes. @RooSoft wdyt? I'll try this solution now.

VajraOfIndra commented:
I modified the structure based on @Stadicus idea. The flow is now:

* Intro
* Preparations: checking SCB location, creating the script with both methods disabled and creating the systemd service
* (Optional) Local backup: guidelines to setup the local backup + enabling the method in the script
* (Optional) Remote backup: guidelines to setup the remote backup + enabling the method in the script
* Test: a final sectio to test the setup (works for local-only, remote-only or both local and remote)

Wdyt?

Stadicus commented:
> GitHub sometimes collapsing comments without any reason

That might have been me, I was already resolving some comments when I've seen your update... I won't do that in the future, sorry.

The current structure looks great. I'll give it one more check in the coming days.

VajraOfIndra commented:
> > GitHub sometimes collapsing comments without any reason
> 
> That might have been me, I was already resolving some comments when I've seen your update... I won't do that in the future, sorry.

If I remember well, they were unresolved comments, so I don;t thin it was you. Moreover, I prefer when reviewers resolves comments once they're happy with the chnages, bevause it makes it easier to see what comment is left to be dealt with!

> The current structure looks great. I'll give it one more check in the coming days.

Ok, great stuff

VajraOfIndra commented:
So I've rewritten the section that explains that the methods in the intoduction.

There is now a new h2 section entitled "Choose your preferred method(s)". It explains what are the two options with a table that summarizes the main differences or PROS/CONS bewteen the two methods. Followed by a paragraph that explains that one or both method can be choosen and how to use the guide.

Here is a preview (can be seen live here: https://vajraofindra.github.io/RaspiBolt/static-channel-backup.html)
![image](https://user-images.githubusercontent.com/86242283/152763228-538b0951-9ffe-4b0d-98fb-af426951932a.png)

The markdown table does not render as beautifully as I thought it would though...

Wdyt of this next section? does it clarify things a bit better while still not favouring one method over the other?

Happy to amend as much you want!

RooSoft commented:
> There is now a new h2 section entitled "Choose your preferred method(s)". It explains what are the two options with a table that summarizes the main differences or PROS/CONS bewteen the two methods. Followed by a paragraph that explains that one or both method can be choosen and how to use the guide.

Love it!!!

RooSoft commented on preparations.md:
```diff
@@ -36,6 +36,7 @@ You need the following hardware:
 * Raspberry Pi 4, with 4+ GB RAM
 * official Raspberry Pi power adapter: other adapters are a common cause for reliability issues
 * external storage: 1+ TB with USB3, an SSD is recommended
+* optional: a small USB thumbdrive or microSD card
```
As a newcomer, I would tend to be curious about what that extra storage might be for...

I still think these options should be enumerated in a separate and more detailed section helping people make informed decisions right from the start.

Moreover, the very next paragraph implies that the node will not boot from the microSD, with the microSD as an optional option in "requirements"... seems a tad confusing to me.

Stadicus replied:
We could add "...thumbdrive or microSD card to store backups" or similar

RooSoft commented on static-channel-backup.md:
```diff
@@ -0,0 +1,450 @@
+| LOCAL       | YES               | NO                      | Drive failure only                     |NO                   | 
+| REMOTE      | NO                | YES                     | Drive failure & widespread node damage |YES                  | 
+
+We recommend to use both methods, but you can choose either one of them, depending on your own requirements and preferences. Wahetever method you've chosen, follow the "Preparations" section, then follow the optional local or/and remote backup sections depending on your choosen method(s). Finally, folllow the "Test" section that works for whatever method you've chosen.
```
Wahetever -> Whatever

folllow -> follow

RooSoft reviewed (changes requested):
Two typos and a potentially confusing preparation section

VajraOfIndra commented:
> Two typos and a potentially confusing preparation section

I fixed the two typos thanks.

I rewritten parts of the hardware requirements to make it less confusing. Let me know what you think of the new wording.

RooSoft reviewed (approved):
Great change!

Stadicus commented on static-channel-backup.md:
```diff
@@ -0,0 +1,450 @@
+
+### Check SCB location
+
+By default, LND saves the SCB file here: `~/.lnd/.lnd/data/chain/bitcoin/mainnet/channel.backup`. Make sure that the `lnd.conf` does not contain the `backupfilepath` option that modifies the backup location (as was used in a previous version of the RaspiBolt v3).
```
Seen from user "lnd" perspective, the path should be:
```
~/.lnd/data/chain/bitcoin/mainnet/channel.backup
```

Maybe it would be better to give the absolute path that's true for every user, including "admin"? 
```
/data/lnd/data/chain/bitcoin/mainnet/channel.backup
```

RooSoft replied:
Is there a specific reason why that is?

Found it confusing at times, as I am trying to document the testnet node... added potential errors, as I could often use the wrong home path... `~` removes this possibility.

Stadicus commented on static-channel-backup.md:
````diff
@@ -0,0 +1,450 @@
+* Reload the systemd service to take in account the change
+
+  ```sh
+  $ systemctl daemon-reload
````
See comment above, we want to use `sudo systemctl restart scb-backup`

Stadicus reviewed (commented):
Alright, I went over the whole text and made "some" wording recommendations. My goal is to keep it concise, in active voice, removing filler words. I also use Grammarly to help me with that. 

I apologize, it looks bad. But it's only recommendations, how **I** would write it. So feel free to take whatever you find useful, and ignore everything else. :)

You can see my changes here:
https://github.com/Stadicus/raspibolt-dev/commit/7967b8c370dba1a436d8da8789595319f4e227bc

VajraOfIndra commented:
> Alright, I went over the whole text and made "some" wording recommendations. My goal is to keep it concise, in active voice, removing filler words. I also use Grammarly to help me with that.
> 
> I apologize, it looks bad. But it's only recommendations, how **I** would write it. But it's **your** guide, so feel free to take whatever you find useful, and ignore everything else. :)
> 
> You can see my suggestions here: [Stadicus@7967b8c](https://github.com/Stadicus/raspibolt-dev/commit/7967b8c370dba1a436d8da8789595319f4e227bc)

Excellent. Very happy to implement all your recommendations. You're the one that has written such a clear and well-written guide so I'm glad to learn how to write more concisely and didactically :) You also removed a lot of the parentheses I was using which is something @RooSoft noted so that's great as I won't have to go through the whole text to reduce their numbers. I've noted I'm addicted to parentheses.. it's a great way to nuance statements.. but not great for readibility I agree!

I managed to merge your branch into mine after resoliving a small conflict. I'm glad I didn't need to do a manual copy/pasting of each paragraph..

I'd be happy with similar wording suggestions in any future PRs from both of you, no need to feel bad about it ;)

Stadicus reviewed (approved):
A few final nits, and then LETS GO TO MASTER!!! :rocket: 

Edit: forgot to submit the comment about the absolute/relative path. I don't have a strong opinion either way. I see @RooSoft's point, le't just leave it as is.

VajraOfIndra commented:
> A few final nits, and then LETS GO TO MASTER!!! rocket
> 
> Edit: forgot to submit the comment about the absolute/relative path. I don't have a strong opinion either way. I see @RooSoft's point, le't just leave it as is.

Great, I implemented your final suggestions and opted for using the relative path for the SCB comment.
PTAL

Stadicus commented on preparations.md:
```diff
@@ -45,6 +44,11 @@ You can buy a cheap hard disk enclosure and reuse an old hard disk.
+You might also want to get this optional hardware:
+
+* optional: Raspberry Pi case to protect your Pi and cool it down
+* optional: optional: a small USB thumbdrive or microSD card to create regular local backups of your Lightning channels
```
* optional: ~~optional:~~ a small USB thumbdrive

Stadicus reviewed (approved):
Final nit? :-P
